File: createTextFile.js
//Hoerpslfolgen-Liste erstellen 
//node createTextFile.js bob

//Libs & Configs
const dayjs = require('dayjs');
const fs = require('fs-extra');
const path = require('path');
const configFile = fs.readJsonSync(path.join(__dirname, "config.json"));
const jsonDir = path.join(configFile.audioDir, "wap", "json");

const hspList = process.argv[2] ? [process.argv[2]] : getHSPList(jsonDir);

//Initialen Liste erstellen
const updateDay = '0000-00-00';

//Update-Liste erstellen
//const updateDay = '2020-11-18';

//Updatedatum fuer Ueberschrift
const updateDayDisplay = (updateDay !== '0000-00-00')
    ? dayjs(updateDay).format('DD.MM.YYYY')
    : dayjs().format('DD.MM.YYYY');

//Ueberschrift oben
console.log("Hörspiel-Liste vom " + updateDayDisplay);

//Ueber einzelne Hoerspiele gehen und Listen erstellen
for (const hsp of hspList) {

    //Daten dieser Serie sammeln und spaeter ausgeben
    const outputData = [];

    //JSON-Datei laden (janosch.json)
    const filePath = path.join(jsonDir, "hsp", hsp + ".json");
    const json = fs.readJSONSync(filePath);

    //Leere Serien (Bibi) uebergehen
    if (!json[0]) {
        continue;
    }

    //Name der Hoerspielserie ausgeben: "Bob der Baumeister - 01 - Bobs Welt" => "Bob der Baumeister " (trim entfernt letztes whitespace)
    const header = (hsp !== "misc") ? (json[0].name).match(/^[^-]+/)[0].trim() : "Sonstige";
    console.log("\n" + header);
    console.log();

    //Ausgabedaten sammeln
    for (const obj of json) {
        if ((!obj.added && updateDay !== '0000-00-00') || obj.added < updateDay) {
            continue;
        }

        //"Wieso Weshalb Warum - Feuerwehr & Polizei" -> "- Feuerwehr & Polizei"
        //"Bob der Baumeister - 32 - Der Spielplatz" -> "- 32 - Der Spielplatz"
        const name = obj.name;
        const name_short = (hsp !== "misc")
            ? (name.match(/-\s*(\d+\s*-\s*.*)$/) || { 0: name })[0].trim()
            : name;

        outputData.push(name_short);
    }

    //Daten dieses HSP sortieren und ausgeben
    outputData.sort((a, b) => {
        // Extract leading numbers for natural sorting
        const numA = parseInt(a.match(/\d+/)?.[0] || 0);
        const numB = parseInt(b.match(/\d+/)?.[0] || 0);

        // If both have numbers, sort numerically
        if (numA && numB) {
            return numA - numB;
        }
        // Otherwise, sort alphabetically
        return a.localeCompare(b);
    });

    console.log(outputData.join("\n"));
}

//Liste der HSP auslesen
function getHSPList(jsonDir) {
    const hspList = [];

    //Hoerspiellisten sammeln
    const files = fs.readdirSync(path.join(jsonDir, "hsp"));
    for (const file of files) {
        hspList.push(file.replace(".json", ""));
    }
    return hspList;
}

File: createTextFile.py
# Hoerpslfolgen-Liste erstellen
# python createTextFile.py bob

import json
import os
import sys
import re
from datetime import datetime
from pathlib import Path

# Load config
with open("config.json", "r") as f:
    config = json.load(f)

json_dir = os.path.join(config["audioDir"], "wap", "json")

# Get HSP list from command-line argument or by reading files
if len(sys.argv) > 1:
    hsp_list = [sys.argv[1]]
else:
    hsp_list = get_hsp_list(json_dir)

# Initial list creation
update_day = '0000-00-00'

# Update list creation
# update_day = '2020-11-18'

# Format date for heading
if update_day != '0000-00-00':
    update_day_display = datetime.strptime(update_day, '%Y-%m-%d').strftime('%d.%m.%Y')
else:
    update_day_display = datetime.now().strftime('%d.%m.%Y')

# Print heading
print(f"Hörspiel-Liste vom {update_day_display}")

# Process each HSP series
for hsp in hsp_list:
    # Collect data for this series
    output_data = []

    # Load JSON file for this series
    file_path = os.path.join(json_dir, "hsp", hsp + ".json")
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            json_data = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        continue

    # Skip empty series
    if not json_data or len(json_data) == 0:
        continue

    # Extract series name: "Bob der Baumeister - 01 - Bobs Welt" => "Bob der Baumeister"
    first_name = json_data[0]["name"]
    if hsp != "misc":
        match = re.match(r'^[^-]+', first_name)
        header = match.group(0).strip() if match else "Unknown"
    else:
        header = "Sonstige"

    print(f"\n{header}")
    print()

    # Collect output data
    for obj in json_data:
        # Filter by update date if applicable
        if (not obj.get("added") and update_day != '0000-00-00') or (obj.get("added", '9999-99-99') < update_day):
            continue

        # Extract short name
        name = obj["name"]
        if hsp != "misc":
            # "Bob der Baumeister - 32 - Der Spielplatz" => "- 32 - Der Spielplatz"
            match = re.search(r'-\s*(\d+\s*-\s*.*)$', name)
            name_short = match.group(0).strip() if match else name
        else:
            name_short = name

        output_data.append(name_short)

    # Sort data by episode number
    def sort_key(entry):
        # Extract leading numbers for natural sorting
        match = re.search(r'\d+', entry)
        num = int(match.group(0)) if match else 0
        
        # Return tuple: (has_number, number, string) for proper sorting
        return (num > 0, num, entry)

    output_data.sort(key=sort_key)

    # Print output
    print("\n".join(output_data))

# List all HSP files
def get_hsp_list(json_dir):
    hsp_list = []
    
    # Collect HSP lists
    hsp_dir = os.path.join(json_dir, "hsp")
    if os.path.exists(hsp_dir):
        for file in os.listdir(hsp_dir):
            if file.endswith(".json"):
                hsp_list.append(file.replace(".json", ""))
    
    return hsp_list
